### ISSUE TITLE
Layout blockGap / spacing: Add support for gap set at the block level in theme.json and global styles

### ISSUE THREAD

**Issue by andrewserong** [WordPress/gutenberg#39789]
Following on from #39601 where we removed the blockGap / block spacing controls in global styles (since the gap value is not currently supported at the block level in global styles), this issue proposes a way to add it back in!

As of https://github.com/WordPress/gutenberg/pull/37360 where the CSS variable approach for `blockGap` was removed, we haven't had support for `blockGap` at the individual block level in global styles or `theme.json`. One possible path forward to re-instating it could be:

* Every block that uses the layout support is server-rendered and currently attempts to grab the block gap value from the block's attributes, and if it doesn't exist, falls back to using the global blockGap CSS variable.
* Can we add in logic in `layout.php` and in the JS rendering for the layout support so that if the block's attributes do contain a `blockGap` setting we then look-up `theme.json` to grab the `blockGap` value if one is set at the corresponding block name level?
* If this approach works, also roll it out to the Gallery block's custom code that implements gap support.

**Comment by ramonjd** [comment]
> As of https://github.com/WordPress/gutenberg/pull/37360 where the CSS variable approach for blockGap was removed, we haven't had support for blockGap at the individual block level in global styles or theme.json.

I've been testing and it looks like it's because `blockGap` is constrained to a `top level style`. I commented about that over in [this PR](https://github.com/WordPress/gutenberg/pull/39845#issuecomment-1081712594), 

Specifically, this [line](https://github.com/WordPress/gutenberg/pull/37360/files#diff-32a8dd7481b8621c19534034f929f125a711de1746d0f97f99bba814a2a1bb77R288):

```php
'spacing' => array(
     ....
    'blockGap' => 'top', // <- this will be skipped when adding to `$schema['styles']['blocks']` 
),
```

There's a comment in `sanitize()` too:

```
// Some styles are only meant to be available at the top-level (e.g.: blockGap),
// hence, the schema for blocks & elements should not have them.
```

> Can we add in logic in layout.php and in the JS rendering for the layout support so that if the block's attributes do contain a blockGap setting we then look-up theme.json to grab the blockGap value if one is set at the corresponding block name level?

Previously, Gutenberg rendered a global style targeting the block class:

```css
.wp-block-group { 
    --wp--style--block-gap: 341px;
 }
```

Just so I understand, would we try to reinstate that rule, or something like that rule? Or look up the global value and only set for the specific block if that block itself hasn't got one (I think the latter?)

**Comment by ramonjd** [comment]
Something like this?

https://github.com/WordPress/gutenberg/pull/39870

**Comment by andrewserong** [comment]
Just a quick drive-by comment while I'm AFK, thank you so much for picking up this issue @ramonjd! The idea in #39870 is exactly the kind of thing I had in mind — looks like a good direction to investigate to me 😀

In principle:

* The global styles only renders out the block gap CSS variable at the root level
* The Layout support is responsible for grabbing block-level values in theme.json and uses those values to compute the right `$gap_value` as you've done in that PR

**Comment by ndiego** [comment]
@andrewserong is this still something we are trying to get resolved in 6.0. If not, I will remove from the [project board](https://github.com/orgs/WordPress/projects/27/views/1). Thanks!

**Comment by ramonjd** [comment]
> @andrewserong is this still something we are trying to get resolved in 6.0. If not, I will remove from the [project board](https://github.com/orgs/WordPress/projects/27/views/1). Thanks!

I've removed this issue and the [PR](https://github.com/WordPress/gutenberg/pull/39870#issuecomment-1103195772). Thanks!

**Comment by andrewserong** [comment]
Implemented in #40875.
